plugins/out_s3/s3.c

```diff
@@ -423,6 +485,67 @@ static int cb_s3_init(struct flb_output_instance *ins,
             flb_plg_error(ctx->ins, "'s3_key_format' must start with a '/'");
             return -1;
         }
+        if (strstr((char *) tmp, "$INDEX")) {
```

Comment:
this is a huge amount of new and complex code for the init function- please move it to a helper function to make things easier to read/cleaner